File: words-api/Lib/BridgeTypes/Domain.cs
// Project: words-api
// File: Domain.cs
// 
// Author: Ethan Kisiel (ethan.a.kisiel@gmail.com)
// 
// File Created: 23 08 2025 23:08:45
// Last Modified: 23 08 2025 23:08:20
// 
// Modified By: Ethan Kisiel (ethan.a.kisiel@gmail.com>)
// 
// Copyright 2025 - 2025 Ethan Kisiel, Ethan Kisiel

namespace words_api.Lib.Enums;

public class Domain
{
    public const string Default = "X";
    public const string Agriculture = "A";
    public const string Medical = "B";
    public const string Arts = "D";
    public const string Religious = "E";
    public const string Scholarly = "G";
    public const string Law = "L";
    public const string Poetry = "P";
    public const string Science = "S";
    public const string Technology = "T";
    public const string Military = "W";
    public const string Mythology = "M";

    // X,      --  All or none
    // A,      --  Agriculture, Flora, Fauna, Land, Equipment, Rural
    // B,      --  Biological, Medical, Body Parts
    // D,      --  Drama, Music, Theater, Art, Painting, Sculpture
    // E,      --  Ecclesiastic, Biblical, Religious
    // G,      --  Grammar, Rhetoric, Logic, Literature, Schools
    // L,      --  Legal, Government, Tax, Financial, Political, Titles
    // P,      --  Poetic
    // S,      --  Science, Philosophy, Mathematics, Units/Measures
    // T,      --  Technical, Architecture, Topography, Surveying
    // W,      --  War, Military, Naval, Ships, Armor
    // Y       --  Mythology
}

File: words-frontend/src/lib/models/codes/Domain.ts
export type DomainValue = 'X' | 'A' | 'B' | 'D' | 'E' | 'G' | 'L' | 'P' | 'S' | 'T' | 'W' | 'M';

export class Domain {

    private static default = 'X';
    private static agriculture = 'A';
    private static medical = 'B';
    private static arts = 'D';
    private static religious = 'E';
    private static scholarly = 'G';
    private static law = 'L';
    private static poetry = 'P';
    private static science = 'S';
    private static technology = 'T';
    private static military = 'W';
    private static mythology = 'M';

    static getLongForm(value: DomainValue, detailed=false): string {
        switch (value) {
            case this.default:
                return (!detailed?
                    'Unknown' :
                    'All or none');
            case this.agriculture:
                return (!detailed?
                    'Agriculture' :
                    'Agriculture, Flora, Fauna, Land, Equipment, Rural');
            case this.medical:
                return (!detailed?
                    'Medical' :
                    'Biological, Medical, Body Parts');
            case this.arts:
                return (!detailed?
                    'Arts' :
                    'Drama, Music, Theater, Art, Painting, Sculpture');
            case this.religious:
                return (!detailed?
                    'Religious' :
                    'Ecclesiastic, Biblical, Religious');
            case this.scholarly:
                return (!detailed?
                    'Scholarly' :
                    'Grammar, Rhetoric, Logic, Literature, Schools');
            case this.law:
                return (!detailed?
                    'Law' :
                    'Legal, Government, Tax, Financial, Political, Titles');
            case this.poetry:
                return (!detailed?
                    'Poetry' :
                    'Poetic');
            case this.science:
                return (!detailed?
                    'Science' :
                    'Science, Philosophy, Mathematics, Units/Measures');
            case this.technology:
                return (!detailed?
                    'Technology' :
                    'Technical, Architecture, Topography, Surveying');
            case this.military:
                return (!detailed?
                    'Military' :
                    'War, Military, Naval, Ships, Armor');
            case this.mythology:
                return (!detailed?
                    'Mythology' :
                    'Mythology');
            default:
                throw new Error('Failed to match domain value');
        }
    }
}